### DISCUSSION TITLE
post/given-test-code/

### DISCUSSION THREAD

**Discussion by giscus** [kakaopay/techlog-comments#91] [Announcements]
# post/given-test-code/

Mock 테스트 코드 작성 중에 마주한 문제들과 그 문제를 해결하는 방법과 노하우를 소개드립니다.

https://tech.kakaopay.com/post/given-test-code/

**Comment by heli-os** [comment]
안녕하세요, 윤! 시리즈 전부 잘 읽었습니다.

두 가지 궁금한게 있는데요.

**1. 본문의 예시와 같이 데이터를 셋업하고, 테스트를 수행하려면 테스트용 데이터베이스를 프로비저닝해야만 할 것 같습니다. 테스트용 데이터베이스는 어떤 방식으로 프로비저닝하여 테스트에 이용하시는지,**

**2. jpa 의존성이 담긴 테스트를 실행하기 위해선 `@SpringBootTest` , `@DataJpaTest` 같은 어노테이션이 필요하고 Spring Context Load가 발생하게 될텐데 단순 메소드의 유닛 테스트도 이렇게 진행하시는지, 다른 방식으로 진행하시는지 궁금합니다 :)**

2-1. 만약 단순 메소드의 유닛 테스트에서 Spring Context를 Load하고 있지 않으시다면, 테스트에 필요한 Data Model을 생성하기 위해 Fixture 등을 관리하여야 할텐데 이 문제는 어떻게 해결하셨나요?

2-2. 만약 단순 메소드의 유닛 테스트에서 Spring Context를 Load하고 있으시다면, 각 테스트의 실행 속도뿐 아니라 테스트 전체의 실행 속도에 큰 영향을 미칠 것 같은데요. 이 문제는 어떻게 해결하셨나요? 그 외에도 여러 테스트간 데이터 격리가 잘 되지 않아 발생하는 데이터 정합성/경합의 문제도 있을 것 같아 여쭈어봅니다.

**Reply by cheese10yun** [reply]
안녕하세요. 답변이 좀 늦었네요. 
> 1. 본문의 예시와 같이 데이터를 셋업하고, 테스트를 수행하려면 테스트용 데이터베이스를 프로비저닝해야만 할 것 같습니다. 테스트용 데이터베이스는 어떤 방식으로 프로비저닝하여 테스트에 이용하시는지,

테스트용 데이터베이스 및 테스트 인프라는 Testcontainers로 관리하고 있습니다. 이와 관련된 Testcontainers 사용 방법에 대한 포스팅입니다. 카카오페이 정산팀에서도 유사한 문제를 겪고, 비슷한 해결 방법을 사용해 이를 공유드립니다: [Testcontainers 관련 포스팅](https://flex.team/blog/2024/07/29/tech-testcontainers/).

> 2. jpa 의존성이 담긴 테스트를 실행하기 위해선 @SpringBootTest , @DataJpaTest 같은 어노테이션이 필요하고 Spring Context Load가 발생하게 될텐데 단순 메소드의 유닛 테스트도 이렇게 진행하시는지, 다른 방식으로 진행하시는지 궁금합니다 :)

JPA 의존성이 필요한 경우 `@SpringBootTest`를 사용하여 테스트하며, 이로 인해 Spring Context Load를 활용합니다. 단순 메소드 유닛 테스트라도 Spring 관련 의존성이 필요한 경우에는 동일하게 진행하고 있습니다.

여담이지만, 개인적으로는 의존성이 필요 없는 POJO 테스트를 선호합니다. 특히 계산식이 복잡한 로직의 경우, POJO 기반으로 로직을 작성하고 관련 테스트 코드도 POJO로 작성하는 것을 선호합니다. 관련 내용은 [스프링캠프 2023](https://youtu.be/XSkz0kO7J3w?si=f7quF2k76W3IOOIM&t=2267)에서도 확인할 수 있습니다.


> 2-1. 만약 단순 메소드의 유닛 테스트에서 Spring Context를 Load하고 있지 않으시다면, 테스트에 필요한 Data Model을 생성하기 위해 Fixture 등을 관리하여야 할텐데 이 문제는 어떻게 해결하셨나요?

Spring Context를 Load하고 있습니다.

> 2-2. 만약 단순 메소드의 유닛 테스트에서 Spring Context를 Load하고 있으시다면, 각 테스트의 실행 속도뿐 아니라 테스트 전체의 실행 속도에 큰 영향을 미칠 것 같은데요. 이 문제는 어떻게 해결하셨나요? 그 외에도 여러 테스트간 데이터 격리가 잘 되지 않아 발생하는 데이터 정합성/경합의 문제도 있을 것 같아 여쭈어봅니다.

**Spring Context를 Load 실행 속도 문제**
Spring Context를 Load할 때의 실행 속도 문제는, Spring Context가 여러 번 로드되면 테스트 코드의 전체 빌드 시간이 길어지기 때문에 발생합니다. 이를 해결하기 위해 Spring Context를 최대한 재사용할 수 있는 방식으로 테스트 코드를 작성하고 있습니다. 관련 내용은 [스프링캠프 2023](https://youtu.be/XSkz0kO7J3w?si=j1Do7O3sQ23Sfcin&t=496)에서도 확인할 수 있습니다.

질문에서 언급하신 `@DataJpaTest`와 같은 Slice Testing 방법도 있지만, 카카오페이 정산팀에서는 `@SpringBootTest`로 로드한 Spring Context를 재사용하는 것이 전체 테스트 빌드 시간에 유리하다고 판단하여, Slice Testing은 사용하지 않고 있습니다. 그 외에도 한 번 로드한 Spring Context를 재사용하기 위한 다양한 방법을 적용하고 있습니다.

**데이터 정합성/경합의 문제**

데이터 정합성 문제의 경우, Spring에서 지원하는 트랜잭션 롤백 전략을 최대한 활용하고 있습니다. 자동 트랜잭션 롤백을 사용할 수 없는 경우에는 수동으로 데이터를 삭제하며, 이를 간편하게 처리할 수 있는 다양한 방법들을 제공하여 진행하고 있습니다.

경합의 문제 경우
경합 문제란 여러 테스트 코드가 동시에 실행되면서 동일한 데이터에 대해 입력, 수정, 삭제 작업을 수행할 때 발생할 수 있는 문제를 의미한다면, 각 모듈(Gradle 모듈) 간의 테스트 인프라를 격리하여, 전체 테스트 빌드를 돌리더라도 모듈 간의 데이터 경합 문제가 발생하지 않도록 관리하고 있습니다.


질문 주셔서 감사합니다. 제 답변이 도움이 되었길 바랍니다.
